CORS-3689: API changes in config/v1 to hold eipAllocations in cluster ingress object.

miheer opened the pull request:
The eipAllocations when set by installer for default AWS NLB IngressController are set in the cluster ingress object the for the ingress operator to refer and then create a default Ingress Controller. This API field holds the eipAllocations value set by installer which were set via install-config.yaml

miheer commented on config/v1/types_ingress.go:
```diff
@@ -152,8 +152,58 @@ type AWSIngressSpec struct {
+	// network load balancer. Present only if type is NLB.
+	//
+	// +optional
+	NetworkLoadBalancerParameters *AWSNetworkLoadBalancerParameters `json:"networkLoadBalancerParameters,omitempty"`
```
yes done

miheer commented:
@JoelSpeed @gcs278 PTAL

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -131,6 +131,7 @@ type LoadBalancer struct {
 
 // AWSIngressSpec holds the desired state of the Ingress for Amazon Web Services infrastructure provider.
 // This only includes fields that can be modified in the cluster.
+// +openshift:validation:FeatureGateAwareXValidation:featureGate=SetEIPForNLBIngressController,rule="self.type == 'NLB' ? true : !has(self.networkLoadBalancerParameters)",message="Network Load Balancer parameters are allowed only when load balancer type is NLB."
```
For consistency with other APIs, can you copy the style of please https://github.com/openshift/api/blob/4f6053f954b069f7fea5788d4307f4213db53678/example/v1/types_stable.go#L124C140-L124C199

miheer replied:
OK done

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -152,8 +153,59 @@ type AWSIngressSpec struct {
+type AWSNetworkLoadBalancerParameters struct {
+	// eipAllocations is a list of IDs for Elastic IP (EIP) addresses that
+	// are assigned to the Network Load Balancer.
+	// The following restrictions apply:
```
I don't think this line adds much to the description of the field

miheer replied:
OK removed it

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -152,8 +153,59 @@ type AWSIngressSpec struct {
+	//
+	// +optional
+	// +listType=atomic
+	// +openshift:validation:FeatureGateAwareXValidation:featureGate=InstallEIPForDefaultNLBIngressController,rule=`self.all(x, self.exists_one(y, x == y))`,message="eipAllocations cannot contain duplicates"
```
Does this need to be gated? You have the ungated version below and I'd expect that to be sufficient

miheer replied:
OK

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -152,8 +153,59 @@ type AWSIngressSpec struct {
+	// are assigned to the Network Load Balancer.
+	// The following restrictions apply:
+	//
+	// eipAllocations can only be used with external scope, not internal.
```
Where is scope defined? We can validate this using CEL right?

miheer replied:
I think I need to remove this. I don't see an API in the installer to set this. Also in the CIO we by default set it to External after installation.

JoelSpeed replied:
But if this is a constraint of the functionality, what would happen if this were set, and the LB was internal?

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -152,8 +153,59 @@ type AWSIngressSpec struct {
+	// The following restrictions apply:
+	//
+	// eipAllocations can only be used with external scope, not internal.
+	// An EIP can be allocated to only a single IngressController.
```
```suggestion
	// Each EIP may only be allocated to a single IngressController.
```

miheer replied:
OK

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -152,8 +153,59 @@ type AWSIngressSpec struct {
+	// The number of EIP allocations must match the number of subnets that are used for the load balancer.
+	// Each EIP allocation must be unique.
+	// A maximum of 10 EIP allocations are permitted.
+	//
```
```suggestion
	// Each EIP allocation must be of the form 'eipalloc-' followed by exactly 17 hexadecimal characters (0-9, a-f, A-F).
	//
```

miheer replied:
OK

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -152,8 +153,59 @@ type AWSIngressSpec struct {
+	// +openshift:validation:FeatureGateAwareXValidation:featureGate=InstallEIPForDefaultNLBIngressController,rule=`self.all(x, self.exists_one(y, x == y))`,message="eipAllocations cannot contain duplicates"
+	// +kubebuilder:validation:XValidation:rule=`self.all(x, self.exists_one(y, x == y))`,message="eipAllocations cannot contain duplicates"
+	// +kubebuilder:validation:MaxItems=10
+	EIPAllocations []EIPAllocation `json:"eipAllocations"`
```
I think this is a workload API so we want this to have omitempty

miheer replied:
OK

JoelSpeed replied:
Review our [documentation](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#configuration-vs-workload-apis). An ingress is installed many times by namespace admins, rather than cluster admins typically right?

JoelSpeed commented on config/v1/tests/ingresses.config.openshift.io/InstallEIPForDefaultNLBIngressController.yaml:
```diff
@@ -0,0 +1,256 @@
+                networkLoadBalancerParameters:
+                  eipAllocations:
+                  - eipalloc-1234567891011abcd
+                  - eipalloc-1234567891011abcf
```
Nit, always finish files with a newline character

miheer replied:
OK

miheer commented:
@JoelSpeed I have made the requested changes. PTAL.

miheer commented:
@JoelSpeed  the tests are failing with the following error:
INSUFFICIENT CI testing for "InstallEIPForDefaultNLBIngressController".
F1022 01:20:04.603312  172416 root.go:64] Error running codegen: error: "install should succeed: overall" only passed 94%, need at least 95% for "InstallEIPForDefaultNLBIngressController" on {azure amd64 ha }
make: *** [Makefile:56: verify-scripts] Error 255

Does it mean I need to add more tests ?

JoelSpeed commented on features/features.go:
```diff
@@ -64,6 +64,13 @@ var (
+								reportProblemsToJiraComponent("Networking / router").
+								contactPerson("miheer").
+								productScope(ocpSpecific).
+								enableIn(configv1.Default, configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
```
Why are you adding this to the default featureset straight away, all features must now go through tech preview first

JoelSpeed reviewed (commented):
I think we need further discussion around the scope comment, how is the scope internal/external related to NLBs here, where in this API do we represent whether an NLB is internal or external?

JoelSpeed commented:
> @JoelSpeed the test ci/prow/verify is failing with the following error:
INSUFFICIENT CI testing for "InstallEIPForDefaultNLBIngressController".
F1022 01:20:04.603312 172416 root.go:64] Error running codegen: error: "install should succeed: overall" only passed 94%, need at least 95% for "InstallEIPForDefaultNLBIngressController" on {azure amd64 ha }
make: *** [Makefile:56: verify-scripts] Error 255
>
> Does it mean I need to add more tests ?

Have you added any tests to openshift/origin that are gated by the feature gate you're introducing here?

gcs278 commented on config/v1/types_ingress.go:
```diff
@@ -131,6 +131,7 @@ type LoadBalancer struct {
 
 // AWSIngressSpec holds the desired state of the Ingress for Amazon Web Services infrastructure provider.
 // This only includes fields that can be modified in the cluster.
+// +openshift:validation:FeatureGateAwareXValidation:featureGate=InstallEIPForDefaultNLBIngressController,rule="self.type == 'NLB' ? true : !has(self.networkLoadBalancerParameters)",message="networkLoadBalancerParameters are forbidden when type is not NLB."
```
Cross link to https://github.com/openshift/enhancements/pull/1634#discussion_r1814173039. I think this CEL is okay for now (and better to keep it here), but I might have to remove it later depending on the designs of the subnets API.

gcs278 reviewed (commented):
Generally looks good to me. I have a lot of open questions that I have explored in https://docs.google.com/document/d/1Y-Z8gaKYv5dczbdwvXanbzBl-HjZ7zasRC0WaJGBWkA/edit?tab=t.0#heading=h.sya755lm2fbe about the IngressConfig API.

The main one that I hope to get answered soon that affects this PR is
["_Should we split Ingress Config into defaulting for the default IngressController and defaulting for user-created IngressControllers?_"](https://docs.google.com/document/d/1Y-Z8gaKYv5dczbdwvXanbzBl-HjZ7zasRC0WaJGBWkA/edit?tab=t.0#heading=h.1196heeu6pjk)

My concern is that IngressConfig.spec.loadBalancer is growing with new fields that have inconsistent purposes. I'd hope to have a discussion with the API team soon about this. @JoelSpeed Maybe the answer is to keep on the path we are (continue with this PR as is) and look for a path for deprecation or a V2.

miheer commented:
> here in this API do we represent whether an NLB is int

Actually we don't set scope via installer. So, that is why I did not feel the need to add a field for this.
The CIO by default  creates a ingress controller with external scope. If a default ingresscontroller is created by the installer with eips the scope will be set to external by default. If someone later tries to change the scope from the operator/v1 ingress i.e Ingress Controller then it won't be allowed.

miheer commented on config/v1/types_ingress.go:
```diff
@@ -152,8 +153,55 @@ type AWSIngressSpec struct {
+	//
+	// +openshift:enable:FeatureGate=InstallEIPForDefaultNLBIngressController
+	// +optional
+	NetworkLoadBalancerParameters *AWSNetworkLoadBalancerParameters `json:"networkLoadBalancerParameters,omitempty"`
```
For consistency I can keep this name.

gcs278 replied:
Sorry, not sure if I understand, were you planning to update to `networkLoadBalancer` for consistency?

miheer commented:
> Generally looks good to me. I have a lot of open questions that I have explored in https://docs.google.com/document/d/1Y-Z8gaKYv5dczbdwvXanbzBl-HjZ7zasRC0WaJGBWkA/edit?tab=t.0#heading=h.sya755lm2fbe about the IngressConfig API.
> 
> The main one that I hope to get answered soon that affects this PR is ["_Should we split Ingress Config into defaulting for the default IngressController and defaulting for user-created IngressControllers?_"](https://docs.google.com/document/d/1Y-Z8gaKYv5dczbdwvXanbzBl-HjZ7zasRC0WaJGBWkA/edit?tab=t.0#heading=h.1196heeu6pjk)
> 
> My concern is that IngressConfig.spec.loadBalancer is growing with new fields that have inconsistent purposes. I'd hope to have a discussion with the API team soon about this. @JoelSpeed Maybe the answer is to keep on the path we are (continue with this PR as is) and look for a path for deprecation or a V2.

Should we split Ingress Config into defaulting for the default IngressController and defaulting for user-created IngressControllers?
[Grant] I think our Ingress Config API could benefit from two structures that accomplish two different things
Naming is going to be really difficult. How do you distinguish between default for default ICs and default for user-created ICs?. spec.defaultIngressControllerDefaults and spec.userCreatedIngressControllerDefaults?
[Joel] We should definitely do this, to make the UX of creating the ingress config much more obvious to the end user, though as mentioned, naming is hard, we can discuss this in a wider forum to come up with something everyone is happy with, ingressControllerDefaults with fields of default and userCreated could be the way, but the default is unfortunate, we’d have to document it well
Reference: https://docs.google.com/document/d/1Y-Z8gaKYv5dczbdwvXanbzBl-HjZ7zasRC0WaJGBWkA/edit?tab=t.0#heading=h.1196heeu6pjk
Answer: @JoelSpeed  @deads2k @gcs278 Shall I split it ? However this will deprecate the old API so do we handle this ?

miheer commented:
> > @JoelSpeed the test ci/prow/verify is failing with the following error:
> > INSUFFICIENT CI testing for "InstallEIPForDefaultNLBIngressController".
> > F1022 01:20:04.603312 172416 root.go:64] Error running codegen: error: "install should succeed: overall" only passed 94%, need at least 95% for "InstallEIPForDefaultNLBIngressController" on {azure amd64 ha }
> > make: *** [Makefile:56: verify-scripts] Error 255
> > Does it mean I need to add more tests ?
> 
> Have you added any tests to openshift/origin that are gated by the feature gate you're introducing here?

No, can you point me exactly where the tests are needed ?

JoelSpeed commented:
> If someone later tries to change the scope from the operator/v1 ingress i.e Ingress Controller then it won't be allowed.

How as an end user, would I change the scope from external to internal?

> No, can you point me exactly where the tests are needed ?

Tests are expected in an appropriate location with [origin](https://github.com/openshift/origin/tree/master/test/extended)

gcs278 commented on config/v1/types_ingress.go:
```diff
@@ -152,8 +153,55 @@ type AWSIngressSpec struct {
+// AWSNetworkLoadBalancerParameters holds configuration parameters for an
+// AWS Network load balancer. For Example: Setting AWS EIPs https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/elastic-ip-addresses-eip.html
+type AWSNetworkLoadBalancerParameters struct {
+	// eipAllocations is a list of IDs for Elastic IP (EIP) addresses that
+	// are assigned to the Network Load Balancer.
```
The Ingress Config API in `spec.loadBalancer` has fields that default for both the `default` IngressController, and/or user-created IngressControllers. @JoelSpeed has previously suggested to split the API for each of these purposes/functions, I agree that would be much more explicit and clear.

However, we can't reasonably expect you to fix all of this now, and if you add a new API field without deprecating and relocating the others, I think the API will become even more confusing. Correct me if you disagree @JoelSpeed. I think we plow ahead with the current design, knowing it's not perfect, but it's hopefully something we can fix later with a V2 or a deprecation/restructuring effort.

All of this to say: we need to document what the purpose of this field is. If we can't have a explicit API structure, it should at least be well documented in the API:

```suggestion
	// eipAllocations is a list of IDs for Elastic IP (EIP) addresses that
	// are assigned to the default IngressController's NLB.
```

gcs278 replied:
Well, actually, I might be overstating the effort here. At this current moment, there is actually only two field (`spec.loadBalancer.platform.aws.type` and `spec.domain`) that set "defaults" or the `default` IngressController values.

But still, I don't quite know exactly what to suggest to structure this API. Maybe I'll get @JoelSpeed's feedback tomorrow in API review office hours about this.

miheer commented:
/close
